### ISSUE TITLE
Logging appenders do not pick up the instrumentation key when set via application insights spring boot starter

### ISSUE THREAD

**Issue by trask** [microsoft/ApplicationInsights-Java#1014]
See #1004 and #880.

Workaround is to set the instrumentation key in the appender configuration when using the application insights spring boot starter (and not using the ApplicationInsight.xml), but opening this issue to look into a how to get rid of this extra configuration step.

**Comment by b4rd** [comment]
I solved this by using [this](https://github.com/qos-ch/logback-extensions/wiki/Spring) extension.
This is how my configuration looks like:
```
@Configuration
public class LogbackConfig {

    @Bean
    static ApplicationContextHolder applicationContextHolder() {
        return new ApplicationContextHolder();
    }

    @Bean
    static LoggerContext loggerContext() {
        return (LoggerContext) LoggerFactory.getILoggerFactory();
    }

    @Bean(initMethod = "start", destroyMethod = "stop")
    @DependsOn("telemetryConfiguration")
    static ApplicationInsightsAppender appInsightsAppender(LoggerContext ctx) {
        ApplicationInsightsAppender appender = new ApplicationInsightsAppender();
        appender.setContext(ctx);
        return appender;
    }
}
```

By delaying the instantiation of the appender, it will pick up the TelemetryConfiguration created and configured by spring.

Would be nice to have a simpler way though :)

**Comment by trask** [comment]
Hi @b4rd, @madhunaidu2468, I haven't been able to reproduce this ☹️, can you look at my repro https://github.com/trask/issue-1014 and give me some pointers how I can make it more similar to the applications where you are having the issue?

E.g.
* Are you running your app as jar or war file?
* If war file, what servlet container are you using (and version)?
* What version of Spring Boot are you using?

Or any instructions you could provide that I can use to repro would be great.

Thanks!

**Comment by b4rd** [comment]
Hi @trask,
You're right. I have integration tests with a mock server for testing that the log messages are actually sent. In this setup my tests started passing with this `DelegatingLogbackAppender`.
Now I switched back to using the ApplicationInsightsAppender directly, and I had to add some `Thread.sleep` to make it work again.

I guess it's because now the traces are stored on the file system while the mock server is not running, and then it needs some time to send those first. Can I disable this storage feature during testing? :confused:

**Comment by trask** [comment]
Hi @b4rd, that may also be due to ApplicationInsights batching, try setting `FlushIntervalInSeconds` to `1` (see https://github.com/microsoft/ApplicationInsights-Java/wiki/ApplicationInsights.XML#channel).

**Comment by b4rd** [comment]
I've discovered today that the current behaviour is still not perfect:

When `ApplicationInsightsAppender` gets instantiated by Logback, the eventually created default `TelemetryConfiguration` will load the heartbeat module with default settings, ignoring the values I have in `application.properties`.

There might be other side effects as well.

**Comment by trask** [comment]
This is resolved in [3.0 Preview](https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-in-process-agent) (everything is collected by the agent now)
